core/include/vecmem/containers/device_vector.ipp

```diff
@@ -0,0 +1,279 @@
+/** VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+// System include(s).
+#include <cassert>
+
+namespace vecmem {
+
+   template< typename TYPE >
+   VECMEM_HOST_AND_DEVICE
+   device_vector< TYPE >::device_vector( device_vector_data< value_type > data )
+   : m_size( data.m_size ), m_ptr( data.m_ptr ) {
+
+   }
+
+   template< typename TYPE >
+   VECMEM_HOST_AND_DEVICE
+   device_vector< TYPE >::device_vector( const device_vector& parent )
+   : m_size( parent.m_size ), m_ptr( parent.m_ptr ) {
+
+   }
+
+   template< typename TYPE >
+   VECMEM_HOST_AND_DEVICE
+   device_vector< TYPE >&
+   device_vector< TYPE >::operator=( const device_vector& rhs ) {
+
+      // Prevent self-assignment.
+      if( this == &rhs ) {
+         return *this;
+      }
+
+      // Copy the other object's payload.
+      m_size = rhs.m_size;
+      m_ptr = rhs.m_ptr;
+
+      // Return a reference to this object.
+      return *this;
+   }
+
+   template< typename TYPE >
+   VECMEM_HOST_AND_DEVICE
+   typename device_vector< TYPE >::reference
+   device_vector< TYPE >::at( size_type pos ) {
+
+      // Check if the index is valid.
+      assert( pos < m_size );
```

Comment:
I worry about this failing silently when `NDEBUG` is set. However, we can't really mirror the behaviour of the standard library because we can't really throw in device code. Explicitly documenting might be key here.

Reply:
I'm open to suggestions... Unfortunately I couldn't come up with anything better. I'm just not aware of anything cleaner for error handling in device code than `assert(...)` statements. (Not even separately for the individual platforms...)

We could possibly put in something that would trigger a memory error right there and then. That could possibly simplify debugging in optimised code a bit more. :thinking: